ipinfo/cmd_download.go

```diff
@@ -0,0 +1,225 @@
+
+Options:
+  General:
+    --token <tok>, -t <tok>
+      use <tok> as API token.
+    --help, -h
+      show help.
+
+Outputs:
+	--compress, -c
+	save the file in compressed format.
+	default: false.
+    --format , -f <mmdb | json | csv>
+    output format of the database file.
+    default: mmdb.
+`, progBase)
+}
+
+func cmdDownload() error {
+	var fTok string
+	var fFmt string
+	var fZip bool
+	var fHelp bool
+
+	pflag.StringVarP(&fTok, "token", "t", "", "the token to use.")
+	pflag.StringVarP(&fFmt, "format", "f", "mmdb", "the output format to use.")
+	pflag.BoolVarP(&fZip, "compress", "c", false, "compressed output.")
+	pflag.BoolVarP(&fHelp, "help", "h", false, "show help.")
+	pflag.Parse()
+
+	args := pflag.Args()[1:]
+	if fHelp || len(args) > 2 || len(args) < 1 {
+		printHelpDownload()
+		return nil
+	}
+
+	var token string
+	if fTok == "" {
+		token = gConfig.Token
+	}
+
+	// require token for download.
+	if token == "" {
+		return errors.New("downloading requires a token; login via `ipinfo login` or pass the `--token` argument")
+	}
+
+	// get download format.
+	var format string
+	switch strings.ToLower(fFmt) {
+	case "mmdb":
+		format = "mmdb"
+	case "csv":
+		format = "csv.gz"
+	case "json":
+		format = "json.gz"
+	default:
+		return errors.New("unknown download format")
+	}
+
+	// download the db.
+	switch strings.ToLower(args[0]) {
+	case "asn":
+		err := downloadDb("asn", format, token, fZip)
+		if err != nil {
+			return err
+		}
+	case "country":
+		err := downloadDb("country", format, token, fZip)
+		if err != nil {
+			return err
+		}
+	case "country-asn":
+		err := downloadDb("country_asn", format, token, fZip)
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("database '%v' is invalid", args[0])
+	}
+
+	return nil
+}
+
+func downloadDb(name, format, token string, zip bool) error {
+	url := fmt.Sprintf("%s%s.%s?token=%s", dbDownloadURL, name, format, token)
+
+	// make API req to download the file.
+	res, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// if output not terminal unzip and write to stdout.
+	if fileInfo, _ := os.Stdout.Stat(); (fileInfo.Mode() & os.ModeCharDevice) == 0 {
+		err := unzipWrite(os.Stdout, res.Body)
+		if err != nil {
+			return err
+		}
+	} else {
+		// get filename.
+		var fileName string
+		if len(pflag.Args()) > 2 {
+			fileName = pflag.Args()[2]
+		} else if zip {
+			if format == "mmdb" {
+				fileName = fmt.Sprintf("%s.%s.gz", name, format)
+			} else {
+				fileName = fmt.Sprintf("%s.%s", name, format)
+			}
+		} else {
+			fileName = strings.TrimRight(fmt.Sprintf("%s.%s", name, format), ".gz")
+		}
+
+		// create file.
+		file, err := os.Create(fileName)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer file.Close()
+
+		// save compressed file.
+		if zip {
+			if format == "mmdb" {
+				writer := gzip.NewWriter(file)
+				defer writer.Close()
+
+				body, err := ioutil.ReadAll(res.Body)
```

Comment:
This isn't a streaming solution; it is storing the full size of the DB in RAM. We need to be streaming it to the writer which will then also stream it to disk/file.

Reply:
@rm-Umar this particular line is still not streaming but the MMDB files are really small so it's probably fine. It won't be if we expose this to other MMDBs.

Reply:
Also `ioutil.Readall` is deprecated in favor of `io.ReadAll`